tests/golang.rs

```diff
@@ -111,7 +111,59 @@ macro_rules! tests {
 
         // https://github.com/golang/go/blob/master/test/chan/fifo.go
         mod fifo {
-            // TODO
+            use super::*;
+
+            const N: i32 = 10;
+
+            // AsynchFifo from the go example.
+            #[test]
+            fn async() {
+                let (tx, rx) = channel::bounded(N as usize);
+                for i in 0..N {
+                    tx.send(i);
+                }
+                for i in 0..N {
+                    if let Some(val) = rx.recv() {
+                        if val != i {
+                            panic!("bad receive");
+                        }
+                    }
+                }
+            }
+
+            fn chain(ch: channel::Receiver<i32>, val: i32, in_c: channel::Receiver<i32>, out: channel::Sender<i32>) {
+                in_c.recv();
+                if let Some(v) = ch.recv() {
+                    if v != val {
+                        panic!(val);
+                    }
+                }
+                out.send(1);
+            }
+
+            // SynchFifo from the go example.
+            #[test]
+            fn sync() {
+                let (ch_s, ch_r) = channel::unbounded();
```

Comment:
This should be `channel::bounded(0)`. The same applies to the next two channel constructors.

Reply:
Oh, I thought the example had them as unbounded? I can switch it though.

Reply:
Go doesn't even have unbounded channels. :)

`make(chan int)` is equivalent to `make(chan int, 0)`.

Reply:
Oh maybe I am confused by the [wording](https://golang.org/pkg/builtin/#make) then:

> If zero, or the size is omitted, the channel is unbuffered.

Is that not the same as the unbounded flavor here? I'm just asking to learn at this point. :)

Reply:
Actually I guess it makes sense considering a Go channel without a size is like a synchronization primitive. Whereas in this crate an unbounded channel can send values indefinitely.

Reply:
Well, the terminology is all mixed up and very confusing. :D

| Channel capacity | Name in Go | Name in `std::sync::mpsc` | Name in `crossbeam-channel` |
|---|---|---|---|
| 0 | unbuffered / synchronous | synchronous with capacity of 0 | bounded with capacity of 0 |
| n | buffered / asynchronous | synchronous with capacity of n | bounded with capacity of n |
| inf | N/A | asynchronous | unbounded |

In `crossbeam-channel`, I decided to avoid words like buffered/unbuffered and synchronous/asynchronous because they often tend to be misunderstood.

Instead, I went with bounded/unbounded, which seemed like the clearest choice of words. What do you think?

Reply:
I think that table is great! At the moment I just didn't translate the correct type in my head. I write Go for my day job so when I see a channel without a buffer I think of it being able to send as many values as needed. But it's also true that there must be a receiver ready. So with that in mind, the table helps make it clear.